File: 82-a-double-cola/get_names.cpp
#include <iostream>
#include <cmath>
#include <array>

using namespace std;

int main()
{
    int n;
    cin >> n;

    int q = 5;
    int i = 1;

    while (n > q)
    {
        n -= q;
        i++;
        q *= 2;
    }

    int l = 1 << (i - 1); // same as (int)pow(2, i-1)
    int idx = (int)((n + l - 1) / l) - 1;

    array<string, 5> names = {"Sheldon", "Leonard", "Penny", "Rajesh", "Howard"};

    cout << names[idx] << endl;
}

File: 82-a-double-cola/get_names.py
n = int(input())

names = ["Sheldon", "Leonard", "Penny", "Rajesh", "Howard"]

i = 1
q = 5

while n > q:
    n -= q
    i += 1
    q *= 2
    # print("i", i, "n", n, "q", q)

l = 2 ** (i - 1)
ii = (n + l - 1) // l - 1
# print("i", i, "n", n, "ii", ii)

print(names[ii])
